src/sdl2/render.rs

```diff
@@ -252,24 +256,25 @@ impl<T> Deref for Canvas<T> {
 /// Its context may be shared with the `TextureCreator`.
 ///
 /// The context will not be dropped until all references of it are out of scope.
-pub struct Canvas<T> {
+pub struct Canvas<T, TC> {
     target: T,
-    context: Rc<RendererContext>,
+    context: Rc<RendererContext<TC>>,
 }
 
 /// Alias for a `Canvas` that was created out of a `Surface`
-pub type SurfaceCanvas<'s> = Canvas<Surface<'s>>;
+pub type SurfaceCanvas<'s> = Canvas<Surface<'s>, SurfaceContext<'s>>;
 
 /// Methods for the `SurfaceCanvas`.
-impl<'s> Canvas<Surface<'s>> {
+impl<'s> Canvas<Surface<'s>, SurfaceContext<'s>> {
```

Comment:
I'm sure that with [associated types](https://doc.rust-lang.org/book/associated-types.html) we can make something work to avoid another generic here. How about 

```rust
trait Target {
    type Context;
}
impl Target for Surface {
    type Context = SurfaceContext
}
// same for Texture and Window
```
, make this trait internal use only (if you can't make it private, at least put a warning in the doc like usual) and change the common part of `impl Canvas<T>` to `impl Canvas<T: Target>`. How does it sound ?